### PR TITLE
re-enable Istio test, other misc. code fixes around endpoint

### PR DESCRIPTION
**Pull request by ianvernon**
Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6467)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

EDIT: green build, testing again to assess stability of test.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8591

The Istio test failed; bookinfo pods get stuck in `Init` state.

Cilium crashed due to deadlock detection getting triggered because `regenerateBPF` was waiting for a long time for proxy ACKS:

```
2018-12-18T01:13:02.232031873Z POTENTIAL DEADLOCK:
2018-12-18T01:13:02.232056866Z Previous place where the lock was grabbed
2018-12-18T01:13:02.232060584Z goroutine 117594 lock 0xc00106a750
2018-12-18T01:13:02.232063521Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:78 lock.(*internalRWMutex).RLock ??? <<<<<
2018-12-18T01:13:02.232066554Z /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:509 endpoint.(*Endpoint).regenerateBPF ???
2018-12-18T01:13:02.232069164Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:309 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.232071633Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
2018-12-18T01:13:02.232074289Z 
2018-12-18T01:13:02.232076826Z Have been trying to lock it again for more than 5m10s
2018-12-18T01:13:02.23207926Z goroutine 117796 lock 0xc00106a750
2018-12-18T01:13:02.232092981Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:78 lock.(*internalRWMutex).RLock ??? <<<<<
2018-12-18T01:13:02.232096194Z /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:509 endpoint.(*Endpoint).regenerateBPF ???
2018-12-18T01:13:02.232098924Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:309 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.232101558Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
2018-12-18T01:13:02.232104015Z 
2018-12-18T01:13:02.244698469Z Here is what goroutine 117594 doing now
2018-12-18T01:13:02.244718129Z goroutine 117594 [select, 5 minutes]:
2018-12-18T01:13:02.244720591Z github.com/cilium/cilium/pkg/completion.(*WaitGroup).Wait(0xc003f4e780, 0x0, 0x0)
2018-12-18T01:13:02.244722614Z 	/go/src/github.com/cilium/cilium/pkg/completion/completion.go:104 +0x195
2018-12-18T01:13:02.244724593Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).WaitForProxyCompletions(0xc0032e22c0, 0xc003f4e780, 0x0, 0x0)
2018-12-18T01:13:02.244726554Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:369 +0x161
2018-12-18T01:13:02.244738925Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc0032e22c0, 0x2f34a60, 0xc00030e780, 0xc00427fc20, 0x0, 0x0, 0x0, 0x0)
2018-12-18T01:13:02.244741221Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:572 +0x502
2018-12-18T01:13:02.24474316Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc0032e22c0, 0x2f34a60, 0xc00030e780, 0xc00427fc20, 0x0, 0x0)
2018-12-18T01:13:02.244745188Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:309 +0x778
2018-12-18T01:13:02.244747098Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate.func1(0xc002e0d4e0, 0xc002cabf10, 0xc0032e22c0, 0x2f34a60, 0xc00030e780)
2018-12-18T01:13:02.244749014Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 +0x238
2018-12-18T01:13:02.24475089Z created by github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate
2018-12-18T01:13:02.244752762Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:409 +0x89
2018-12-18T01:13:02.24475462Z Other goroutines holding locks:
2018-12-18T01:13:02.244756495Z goroutine 117594 lock 0xc0032e24a8
2018-12-18T01:13:02.244758306Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:99 lock.(*internalMutex).Lock ??? <<<<<
2018-12-18T01:13:02.244760402Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:235 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.244762283Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
2018-12-18T01:13:02.244764184Z 
2018-12-18T01:13:02.24476601Z goroutine 117594 lock 0xc003f4e798
2018-12-18T01:13:02.2448152Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:99 lock.(*internalMutex).Lock ??? <<<<<
2018-12-18T01:13:02.244826693Z /go/src/github.com/cilium/cilium/pkg/completion/completion.go:98 completion.(*WaitGroup).Wait ???
2018-12-18T01:13:02.244838258Z /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:369 endpoint.(*Endpoint).WaitForProxyCompletions ???
2018-12-18T01:13:02.244840529Z /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:572 endpoint.(*Endpoint).regenerateBPF ???
2018-12-18T01:13:02.244842436Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:309 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.244844351Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
2018-12-18T01:13:02.244846219Z 
2018-12-18T01:13:02.244848046Z goroutine 117861 lock 0xc0032e2a28
2018-12-18T01:13:02.244849841Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:99 lock.(*internalMutex).Lock ??? <<<<<
2018-12-18T01:13:02.244851844Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:235 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.244853751Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
2018-12-18T01:13:02.244855634Z 
2018-12-18T01:13:02.244857404Z goroutine 117796 lock 0xc0032e2ce8
2018-12-18T01:13:02.24485917Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:99 lock.(*internalMutex).Lock ??? <<<<<
2018-12-18T01:13:02.244861165Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:235 endpoint.(*Endpoint).regenerate ???
2018-12-18T01:13:02.244863042Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:435 endpoint.(*Endpoint).Regenerate.func1 ???
```

@jrajahalme is there anything I can look at here to determine what is happening on the Envoy side? I suspect that the VMs get bogged down due to load, but in the prior build, the test ran just fine (and succeeded in a short amount of time), so I'm wondering if there's something else at play.

The cilium pod with the "deadlock" was restarted, but it was restarted after the test failed at `T01:12:51Z`, so it's only relevant to look at the logs before the test failed in `pod-kube-system-cilium-sd6h4-cilium-agent-previous.log`.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Load on the node is still quite high:

```
cmd: "kubectl exec -n kube-system cilium-sd6h4 -- uptime" exitCode: 0 duration: 300.025419ms stdout:
 01:12:21 up 40 min,  0 users,  load average: 9.96, 10.11, 6.22
```

**Comment by ianvernon** [comment]
I think the underlying issue is that we do not have a good way of handling high-load scenarios when compiling programs. We try to use all available CPUs to compile, but if the system is running under high load, I think we should reduce the number of CPUs available. Under high enough load, operations will take longer and longer, especially compilations.

**Comment by ianvernon** [comment]
Build failure: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8595/

Unit test failure unrelated to my changes.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8596/

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
When running the test locally, I observe that the endpoint keeps getting deleted / re-added while the corresponding pod(s) are in the `Init` state. For whatever reason, there is very little information about what happens to these endpoints, which are continuously added and re-added in perpetuity. 

Example in my local environment:

```
# cilium endpoint list
level=info msg="Detected MTU 1500" subsys=mtu
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                                    IPv6                 IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                                                    
...                                                                                                    
15601      Disabled           Disabled          32926      k8s:app=ratings                                                                f00d::a0a:0:0:b8e2   10.10.0.51    regenerating   
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                                                         
                                                           k8s:io.cilium.k8s.policy.istiosidecarproxy=true                                                                                  
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                                  
                                                           k8s:io.kubernetes.pod.namespace=default                                                                                          
                                                           k8s:version=v1                                                                                                                   
                                                           k8s:zgroup=bookinfo                                                                                                               
```

```
# cilium endpoint list
level=info msg="Detected MTU 1500" subsys=mtu
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                                    IPv6                 IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                                                    
...                                                                                
64625      Disabled           Disabled          32926      k8s:app=ratings                                                                f00d::a0a:0:0:b0ef   10.10.0.118   regenerating   
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                                                         
                                                           k8s:io.cilium.k8s.policy.istiosidecarproxy=true                                                                                  
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                                  
                                                           k8s:io.kubernetes.pod.namespace=default                                                                                          
                                                           k8s:version=v1                                                                                                                   
                                                           k8s:zgroup=bookinfo                                     
```

EDIT: what happens is that the CNI ADD fails because the endpoint doesn't regenerate in time, so it gets deleted and a new one is added with the same labels. However, a lot of the logs for endpoints that are added have endpoint ID `0`, which I don't quite understand, because the endpoint is added into the endpoint manager before it is regenerated, and thus is given an ID. I'm working on adding more logs and running the test locally to get more information about the endpoints / the state they are in.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Finally figured out the root cause of why pods were never going out of `Init` state. The problem lies with the fact that when we insert an endpoint into the EndpointManager, we were waiting synchronously for it to be regenerated:

https://github.com/cilium/cilium/blob/5c69b1c611c401c274f359cdd7b1a4d0b99600a5/pkg/endpointmanager/manager.go#L401-L406

This caused a sort of 'deadlock' in the Endpoint creation logic, as regeneration of endpoints involves waiting for the proxy to acknowledged that it got configured with the policy imported into the Cilium agent. However, waits for proxy acknowledgements are not waited for if no L7 rules select the endpoint which is being regenerated, which is why I noticed some pods always got into `Ready` state (productpage, which had no L7 policy in the test), but others didn't (details and ratings, which did have L7 policy imported before the pods were started). The reason this issue revealed itself is because recently, the init identity phase was removed from Cilium, and previously, the policy for the init identity only applied at L3 / L4 in the CI, so there was never any deadlock waiting for ACKs from the proxy. However, if the init policy had any L7 policy, I believe this same issue would have occurred.

I thought about whether we could remove the need to synchronously wait for regeneration upon insertion into the `EndpointManager`, but I'm wary of how this will change the semantics behind returning success in the API when an endpoint is created. What should "successful" creation of an endpoint be? I'd appreciate feedback w.r.t. that. Also, one thing to note is that we do synchronously wait for Endpoints to be regenerated in `createEndpoint` in the CNI case (e.g. Kubernetes and Istio, not Docker plugin case) but instead of using `RegenerateWait`, we wait for a signal to show that the BPF program has been compiled / loaded in the datapath:

https://github.com/cilium/cilium/blob/5c69b1c611c401c274f359cdd7b1a4d0b99600a5/daemon/endpoint.go#L273-L279

However, at this point, the proxy will not be configured, but that is acceptable since the sidecar hasn't even been started yet. Once we return from the API call, pods will go into `ContainerCreating` state, and the istio-sidecar will be started. The regeneration that was triggered asynchronously during the API call will eventually be configured once it starts, and the endpoints will go into `Ready` state in cilium. I'm not sure how this will affect the state of pods - will a pod in `Ready` state in Kubernetes be guaranteed to have its proxy configured by Cilium as well?

Another note - If we decide to not wait synchronously for endpoints to be regenerated any more, then maybe in the NPDS logic, we no longer have to not wait for policy to be configured in Envoy if it does not apply at L7? cc: @jrajahalme

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Green build of just the Istio test: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8733/

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8734

I suspect asynchronous regeneration breaks the CI, which uses the Docker plugin. I can check whether the entire regeneration of endpoints is required (including proxy updates) in `createEndpoint`, and pass this as a parameter to `AddEndpoint` into endpointmanager to preserve the behavior that exists now.

**Comment by ianvernon** [comment]
test-me-please

**Comment by tgraf** [comment]
@ianvernon The RegenerateWait() seems to be an optimization and is conditional on the state. I think we can simply remove it.

**Comment by ianvernon** [comment]
> The RegenerateWait() seems to be an optimization and is conditional on the state. I think we can simply remove it.

I suspect that we will encounter odd behavior in the CI by removing it outright, as both PUT / PATCH operations rely on synchronous regeneration (this happened here: 
https://github.com/cilium/cilium/pull/6467#issuecomment-451313356 ). I think that in the endpoint creation case (PUT) we are always in the expected state. Would you be okay with me merging this change as I have it (provided the tests pass) and focus on removing that separately so that we can get the Istio test re-enabled? I suspect the change you suggested will have a lot of side effects. I can try it out locally.

**Comment by ianvernon** [comment]
Hm, same failure occurred again... https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8735/testReport/(root)/runtime/RuntimePolicyEnforcement_Policy_Enforcement_Always_Container_creation/

Checking locally.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failures: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8736/execution/node/96/log/?consoleFull

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Found cause of build failures, will fix and re-run CI.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go [hunk 1]
```diff
@@ -422,22 +422,21 @@ func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMe
...
 		e.RUnlock()
-		scopedLog := e.getLogger()
 
 		// We should only queue the request after we use all the endpoint's
 		// lock/unlock. Otherwise this can get a deadlock if the endpoint is
 		// being deleted at the same time. More info PR-1777.
 		doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
 		if doneFunc != nil {
-			scopedLog.Debug("Dequeued endpoint from build queue")
+			e.getLogger().Debug("Dequeued endpoint from build queue")
```
Comment (aanm) [inline pkg/endpoint/policy.go:431]: Please leave it as it was before, there's no need to call atomic.Load 3 times.

File: daemon/endpoint.go [hunk 2]
```diff
@@ -231,7 +231,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 		return PutEndpointIDCreatedCode, nil
 	}
 
-	log.Debug("Synchronously waiting for endpoint to regenerate")
+	ep.Logger(daemonSubsys).Debug("Synchronously waiting for endpoint to regenerate")
```
Comment (aanm) [inline daemon/endpoint.go:272]: These changes seem out of scope of commit message

File: daemon/daemon_main.go [hunk 3]
```diff
@@ -87,11 +87,12 @@ const (
...
 	argDebugVerboseEnvoy   = "envoy"
 
-	apiTimeout = 60 * time.Second
+	apiTimeout   = 60 * time.Second
+	daemonSubsys = "daemon"
 )
 
 var (
-	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "daemon")
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, daemonSubsys)
```
Comment (aanm) [inline daemon/daemon_main.go:92]: These changes seem out of scope of commit message

File: pkg/endpointmanager/manager.go [hunk 4]
```diff
@@ -79,8 +79,14 @@ func Insert(ep *endpoint.Endpoint) error {
 		if id == uint16(0) {
 			return fmt.Errorf("no more endpoint IDs available")
 		}
-
 		ep.ID = id
+
+		// Update endpoint logger to contain ID.
+		ep.UnconditionalRLock()
+		ep.UpdateLogger(map[string]interface{}{
+			logfields.EndpointID: ep.ID,
```
Comment (aanm) [inline pkg/endpointmanager/manager.go:85]: ep.ID doesn't need to be locked

File: pkg/endpointmanager/manager.go [hunk 5]
```diff
@@ -399,10 +405,12 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
 	}
 
 	if build {
-		if err := ep.RegenerateWait(owner, reason); err != nil {
-			Remove(ep)
-			return err
-		}
+		go func() {
+			if err := ep.RegenerateWait(owner, reason); err != nil {
```
Comment (aanm) [inline pkg/endpointmanager/manager.go]: If this is inside a go routine how will we return to the caller the regeneration has failed?

File: pkg/endpointmanager/manager.go [hunk 6]
```diff
@@ -405,12 +405,20 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
 	}
 
 	if build {
-		go func() {
+		if asynchronouslyBuild {
```
Comment (aanm) [inline pkg/endpointmanager/manager.go]: I don't see a good reason to add the logic of building the endpoint async or not into the endpoint manager. Besides, the endpoint itself offers a function to perform the regeneration async, `ep.Regenerate`.

File: pkg/endpointmanager/manager.go [hunk 7]
```diff
@@ -405,12 +405,20 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
...
+			go func() {
+				if err := ep.RegenerateWait(owner, reason); err != nil {
+					ep.Logger("endpointmanager").Error("removing endpoint because initial build failed")
+					Remove(ep)
```
Comment (aanm) [inline pkg/endpointmanager/manager.go]: It seems odd to me to regenerate an endpoint and if something fails we remove it from the endpointmanager without having the caller of this function to be aware of that.

**Review by aanm — DISMISSED** [review]

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (tgraf) [reply pkg/endpoint/policy.go:431]: I think at least before and after `regenerate()`, the scopedLog must be refreshed as it will have changed significantly. We should optimize for correctness in the logs instead of micro-optimizing this code at this level. If we want to optimize then let's ensure that we don't call `getLogger()` when debugging is not enabled.

File: pkg/endpointmanager/manager.go (hunk 5 above)
Comment (tgraf) [reply pkg/endpointmanager/manager.go]: Callers could not rely on this as it was dependent on the state already. We only have two callers to `AddEndpoint()`:

1. `createEndpoint()`: When new endpoints are inserted
2. `LaunchAsEndpoint()` when the health endpoint is created.

We should move to `RegenerateWait()` into the callers:

* For `LaunchAsEndpoint()` it should run as a controller to retry on error and also have an interval for rebuilds such as 15 minutes so we have a fallback

* For `createEndpoint()` it should also start a controller to keep retrying to build the endpoint. We can do an interval of 15 minutes as well. When `createEndpoint()` is in `SyncBuildEndpoint` mode then it should first do the sync build and then start the controller.

The controller must be removed on endpoint leave.

Once we have that we can follow-up and use the controller for all endpoint regenerations and we always have retries on build failures.

File: pkg/endpointmanager/manager.go (hunk 6 above)
Comment (tgraf) [inline pkg/endpointmanager/manager.go]: So it seems that for the docker plugin, we have always been waiting for the initial build to complete. It would be better then for the docker plugin to also pass in `SyncBuildEndpoint`. The docker plugin version is tied to the agent version 1:1 so as we change the behavior of `createEndpoint()`, it is fine to require changing the docker plugin as well. In particular as an API user would expect an async build if `SyncBuildEndpoint` is *not* passed in.

File: test/runtime/Policies.go [hunk 8]
```diff
@@ -158,6 +158,8 @@ var _ = Describe("RuntimePolicyEnforcement", func() {
 
 			By("Create a new container")
 			vm.ContainerCreate("new", helpers.HttpdImage, helpers.CiliumDockerNetwork, "-l id.new")
+			areEndpointsReady := vm.WaitEndpointsReady()
```
Comment (tgraf) [inline test/runtime/Policies.go:161]: Please add a statement to the commit message why this is required.

**Review by tgraf — CHANGES_REQUESTED** [review]
Fantastic work. Really glad to see that we got it working without init policies.

File: pkg/endpointmanager/manager.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/endpointmanager/manager.go]: +1 for the retries on build failures!

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/endpoint/policy.go:431]: Nit-picking here, but `getLogger()` is called on average about 1 time only. On monitorAPI failure it's 2 times, but there is no need to optimize that.

**Review by jrajahalme — APPROVED** [review]
Great work! Nothing to add to the thorough reviews by @aanm and @tgraf.

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (ianvernon) [reply pkg/endpoint/policy.go:431]: The reason I did this is because there is a lot of change in state to the endpoint during this process, and I found that information about the endpoint was stale in the log (e.g., endpoint ID was zero at certain points in time even after it had been added to the endpoint manager and allocated an ID). I think the only time we can safely use a scoped log is if the endpoint is locked, and the endpoint structure isn't locked here.

File: daemon/endpoint.go (hunk 2 above)
Comment (ianvernon) [reply daemon/endpoint.go:272]: I'll update the commit message explaining what I've done here.

File: daemon/daemon_main.go (hunk 3 above)
Comment (ianvernon) [reply daemon/daemon_main.go:92]: I'll update the commit message explaining what I've done here.

File: pkg/endpointmanager/manager.go (hunk 4 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go:85]: Ah, did not realize! I will add a comment to the ID field in the Endpoint to note this. Will update accordingly.

File: pkg/endpointmanager/manager.go (hunk 6 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go]: I did this to maintain existing behavior on the Docker plugin side, so I would be reducing the amount of changes introduced in this PR. I also kept the `RegenerateWait` call in so we would remove the endpoint in the async case if the first build failed, as was being done before since I didn't want to change existing behavior too much in that case as well.

File: pkg/endpointmanager/manager.go (hunk 7 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go]: Yeah, I was a bit unsure about what to do here, so I erred on the side of caution by keeping the existing behavior of removing the endpoint from the EndpointManager if the first build for it failed :/

File: pkg/endpointmanager/manager.go (hunk 4 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go:85]: Hm, but `UpdateLogger` documentation says to hold the endpoint's mutex for reading? Should that statement be removed from its documentation?

File: pkg/endpointmanager/manager.go (hunk 5 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go]: I wasn't sure whether that was in the scope of this PR or not. I wanted to minimize the existing changes in behavior in the cilium-agent during one of the most critical parts of code (creation of endpoints) here, so I made the regeneration of the endpoint asynchronous, but still removed it upon failure if the first build failed, which preserves existing behavior as much as possible. I did this because this needs to be backported to the v1.4 branch. @tgraf do you want this change to be made as part of this PR? I'm worried that we may change behavior in a way we don't expect with that sort of a change, which may be better to do as part of the work for v1.5. I may be overly cautious here, though.

**Comment by ianvernon** [comment]
test-me-please

File: test/runtime/Policies.go (hunk 8 above)
Comment (ianvernon) [reply test/runtime/Policies.go:161]: Fixed in latest upload.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: daemon/daemon_main.go (hunk 3 above)
Comment (ianvernon) [reply daemon/daemon_main.go:92]: @aanm updated commit message.

File: daemon/endpoint.go [hunk 9]
```diff
@@ -211,19 +211,37 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
...
 
-	if err := endpointmanager.AddEndpoint(d, ep, "Create endpoint from API PUT"); err != nil {
-		log.WithError(err).Warn("Aborting endpoint join")
-		return PutEndpointIDFailedCode, err
+	if err := ep.LockAlive(); err != nil {
+		d.deleteEndpoint(ep)
+		return PutEndpointIDFailedCode, fmt.Errorf("endpoint was deleted while waiting for the identity")
+	}
+
+	build := ep.GetStateLocked() == endpoint.StateReady
```
Comment (aanm) [inline daemon/endpoint.go]: I don't understand this line, why is the endpoint only being built if it's ready? AFAIK the CNI plugin sends the endpoint to the cilium-agent in the `endpoint.StateWaitingToRegenerate` state

File: daemon/endpoint.go [hunk 10]
```diff
@@ -238,10 +238,21 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 	ep.Unlock()
 
 	if build {
-		if err := ep.RegenerateWait(d, "Initial build on endpoint creation"); err != nil {
-			d.deleteEndpoint(ep)
-			return PutEndpointIDFailedCode, fmt.Errorf("failed to build the endpoint: %s", err)
-		}
+		// Do not synchronously regenerate the endpoint when adding it into the
+		// endpointmanager. We have custom logic later for waiting for specific
```
Comment (aanm) [inline daemon/endpoint.go]: nit, this is not endpointmanager.

File: pkg/endpointmanager/manager.go (hunk 4 above)
Comment (aanm) [reply pkg/endpointmanager/manager.go:85]: @ianvernon ah, I see. The comment should say: `if fields is nil you must hold Endpoint.Mutex for reading.`

**Comment by ianvernon** [comment]
Build failure: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8923/

Appears related to https://github.com/cilium/cilium/issues/6600 . Trying again to see if it persists. FWIW I did not see this behavior before pulling in #6586 .

File: daemon/endpoint.go (hunk 9 above)
Comment (ianvernon) [reply daemon/endpoint.go]: cc: @tgraf  - I cherry-picked this commit per his recommendation. @tgraf can you provide some context here? I can add it as a comment for future reference.

File: daemon/endpoint.go (hunk 10 above)
Comment (ianvernon) [reply daemon/endpoint.go]: Will be fixed in next upload.

File: pkg/endpointmanager/manager.go (hunk 4 above)
Comment (ianvernon) [reply pkg/endpointmanager/manager.go:85]: Ah, got it! Will remove the `RLock` and `RUnlock` in next upload.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

EDIT: got a green build, but rebased and dropped the log message commit that I added.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
This needs to be backported to all branches to which https://github.com/cilium/cilium/pull/6299 was backported because the init identity removal means that any pod which is started in conduit with Istio which has an L7 policy selecting it will get stuck in this 'deadlock' state. I can do this backport since it may cause issues with the CI, and I am an expert at debugging the Istio test by now :)

**Comment by ianvernon** [comment]
test-me-please

File: cilium-health/launch/endpoint.go [hunk 11]
```diff
@@ -266,10 +266,15 @@ func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressin
 		return fmt.Errorf("Error while adding endpoint: %s", err)
 	}
 
+	if err := ep.LockAlive(); err != nil {
```
Comment (tgraf) [inline cilium-health/launch/endpoint.go]: Please add a `Fixes: ` tag to the commit message

**Review by tgraf — CHANGES_REQUESTED** [review]
Please separate out the fixes, in particular, the last commit fixing up the missing lock in the health daemon so we can backport. I doubt that we want to backport all commits of this PR into all branches.

File: cilium-health/launch/endpoint.go (hunk 11 above)
Comment (ianvernon) [reply cilium-health/launch/endpoint.go]: Done and posted as a separate PR. I'd still like this backported to v1.4 because I don't want to break the Istio integration again.

**Comment by ianvernon** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

File: daemon/endpoint.go (hunk 9 above)
Comment (ianvernon) [reply daemon/endpoint.go]: This change is no longer part of this PR so marking as resolved.

**Comment by ianvernon** [comment]
@cilium/ci @cilium/endpoint please review.

**Review by raybejjani — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
running the tests one more time with k8s to catch any possible flake

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-K8s/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
Failed while cilium was restarting
Expected command: kubectl exec -n default netperf-client -- netperf -l 300 -t TCP_STREAM -H 10.10.0.101 
To succeed, but it failed:
Exitcode: 1 
Stdout:
 	 MIGRATED TCP STREAM TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 10.10.0.101 (10.10.0) port 0 AF_INET
	 send_data: data send error: Connection reset by peer (errno 104)
	 netperf: send_omni: send_data failed: Connection reset by peer
	 command terminated with exit code 1
	 
Stderr:
 	 

/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-K8s/src/github.com/cilium/cilium/test/k8sT/Chaos.go:237
```

**Review by eloycoto — APPROVED** [review]

**Comment by ianvernon** [comment]
The above failure looks unrelated to my changes, but I'll trigger the tests again anyway.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
K8s CI passed this time, while normal CI tests did not pass due to Vagrant provisioning errors. Given that the normal CI tests have passed before without failure apart from infrastructural issues, and that the K8s CI passed all tests, I am merging.
